### DISCUSSION TITLE
Full Support for USB Retro Joysticks and Joystick Adapters

### DISCUSSION THREAD

**Discussion by PARALAX1974** [BlitterStudio/amiberry#1296] [Q&A] [answered]
I have a self-made USB joystick adapter for connecting two original C64/Amiga/Atari joysticks, which I ordered from this site: https://restore-store.de/home/58-usb-joystickadapter-donaldsinchai-4000000000587.html
Unfortunately, it is apparently not supported by Amiberry, at least not on Raspberry Pi 4b, although the actual recognition seems to work. In the list of recognized input devices, the adapter is recognized as "hexagons.de Retro-Joystick". This should be identical to the "Retro Donald/Sinchai" version I use. However, I am not 100% sure about that, because the Firmware could be different. In any case, the adapter still worked at least partially with Amiberry 5.2, albeit incorrectly. Steering to the right was recognized as "right-down", steering to the left was recognized as "left-up". The second joystick is also not working. However, in the version 5.6.6 of Amiberry that I tested last night (I can't update to 5.7 yet, as it's not available via Retropie Update Script yet), the adapter no longer works at all, although the inputs in the menu are recognized at least with the "Remap" function in the "Input" menu. However, this type of recognition is apparently only designed for joypads, not for USB Competition Pros or USB joystick adapters like mine, so all joypad buttons are checked. Even individually setting the joystick assignment under "Custom Controls" (which unfortunately is not designed for classic joysticks at all, only modern joypads) did not improve the situation. Not a single button is recognized during emulation. It would be nice if there was support for classic joysticks and my adapter in the future or even in the next Amiberry version.

**Comment by giantclambake** [comment]
The link provided points to a sticker ~ did you mean this? -> https://restore-store.de/home/178-usb-joystickadapter-by-hexagon.html

I'm guessing so, as it identifies as  "hexagons.de Retro-Joystick".

I can't find this in gamecontrollerdb.txt so it must be a Retroarch mapping.

FWIW that firmware source code is here -> https://github.com/hongospongo/joyadapter

(only commenting here due to similarities with problematic 'Manta' gamepad...ie; works in GUI, not in emulation ;)

**Comment by PARALAX1974** [comment]
Yes, Sorry! I have corrected my link above, but I think the firmware is different to the Hexagon due to legal issues. But you or one of the authors of Amiberry could compare the source code to be safe, if you plan support for this hardware. :-)

**Comment by giantclambake** [comment]
No, I'm just a supporter/user of amiberry, who tries to help where I can ;)

I'm unsure how retroarch mappings apply, so maybe there's an issue there ...but what piqued my interest was the device seemingly working in the GUI, but not in emulation ; I have a similar issue with a certain SFC gamepad and just one specific button.  You can have a read of #1237 if you like, it'll give you some idea of how to test things using sdl2-jstest  https://github.com/Grumbel/sdl-jstest and the AmigaTestKit https://github.com/keirf/Amiga-Stuff/releases to get a better idea of what's going on at the SDL2 input layer, and what the test kit reports in the emulation. You should probably use the default mode first, then change to joystick mode in the Inputs panel dropdown for Port 1:  and retest, just to see if anything changes. 

Once you know how sdl2-jstest 'sees' the axis/button events, it gives a reference point. You'll note at the end of that ticket, I did try my basic saitek USB joystick and it worked fine, so it seems to work for (analog, 2button) joystick.

**Comment by midwan** [comment, marked as answer]
This is most likely related to the retroarch mapping for this controller. Since you mentioned you're testing this under RetroPie, which uses Retroarch mappings. Those always take priority in Amiberry and the internal mappings are skipped.

The Remap feature is only applicable on non-Retroarch mappings, so you wouldn't see a difference while those are in use.

You could try disabling the retroarch mapping to see if that helps in your case. Amiberry does a simple check to see if it should use that, by looking for the following things:
- A `retroarch.cfg` file, in the configured location (see the Paths panel)
- A controller mapping file, with the exact Joystick name as it was detected by RetroPie. For example, if your controller is named `XBox 360 controller`, it will look for a file named `XBox 360 controller.cfg`.

If you make sure these two are not found, then Amiberry will automatically fall back to the default controller mappings, or a custom one if you used the Remap feature (which saves to the controllerdb_user.txt file).

Try changing the configured path for the Retroarch.cfg file, and rename the controller `.cfg` file so it's not found (e.g. add a `.bak` at the end of it), then restart Amiberry and check if it's any better.

**Comment by PARALAX1974** [comment]
I finally had time today to take a closer look at it and I would like to say a big thank you for the tip that solved my problem. :-)
I simply changed both lines in the "amiberry.conf" file as follows:

controllers_path=/opt/retropie/emulators/amiberry/no-controllers/
retroarch_config=/opt/retropie/emulators/amiberry/conf/retroarch.bak

After that, the joystick adapter and the two Atari joysticks could be addressed correctly and even via direct detection, just like before. I only need to select the entry "hexagons.de Retro-Joystick" for Port 0 and Port 1 in the Amiberry Config Menu. I didn't even need the remap function for this. Since the adapter is listed twice due to its two ports, the handling is exactly the same as with WinUAE on the PC. To detect it, I only needed to press the fire button (I think this would have happened automatically before due to the Retroarch configuration file). Alternatively, I can also use my two Logitech F310 gamepads without having to restart Amiberry. The only difference is that the fire button is now on the A button instead of the B button, but this is probably because the configuration file can no longer be found.

**Comment by PARALAX1974** [comment]
It's a shame that after changing the above configuration you can no longer use the menu button on the joypad or any of the settings. This now only works with F12 or with the keyboard and mouse. At least I can now use both joysticks on the USB joystick adapter without any problems, but I no longer have the advantages of joypad control (I use a Logitech F310 and the USB adapter with the two Atari joysticks). And since I don't want to change the Retroarch configuration because of this, it will just have to stay like this. There doesn't seem to be a way to subsequently integrate the joypad and not the joysticks on the USB adapter for menu control via the Amiberry menu. It's probably far too complicated.

**Reply by giantclambake** [reply]
This might be related to #1363 in some way, and as midwan says there, it needs to be tested further with different controller hardware. To do that (currently), you'd need compile a binary from the `Master` tree to suit your platform/OS version.
